### PR TITLE
DM-33768: Allow multiple targets for tag/push

### PR DESCRIPTION
**Pull request by athornton**

### CODE REVIEW

**Comment by athornton** [comment]
This feels sort of gross, but I first tried it splitting the string in Makefile syntax and then using make conditionals, and that was even worse.

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -64,41 +64,81 @@ jobs:
...
-             *)
-                 d_registry=${c_image_host}
-                 d_username=''
-                 d_password=''
-                 ;;
-         esac
-         echo "d_registry=${d_registry}" >> $GITHUB_ENV
-         echo "d_username=${d_username}" >> $GITHUB_ENV
-         echo "d_password=${d_password}" >> $GITHUB_ENV
+         split_images=$(echo ${{ env.image}} | tr ',' ' '
```
Comment (timj) [inline .github/workflows/build.yaml:67]: Can you add a comment here that you are creating per-service variables from a comma-separated list of images?

Also inconsistent spacing around `env.image`...

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -64,41 +64,81 @@ jobs:
...
+                     d_registry="ghcr.io"
+                     d_username="token"
+                     d_password=${{ secrets.GHCR_PUSH_TOKEN}}
+                     ;;
+                 *-docker.pkg.dev)
+                     d_tag="google_"
+                     d_registry=${image_host}
+                     d_username="_json_key"
+                     d_password=${{ secrets.GAR_PUSH_TOKEN }}
+                 *)
```
Comment (timj) [inline .github/workflows/build.yaml:96]: Does this fall through ever trigger? Should there be an `echo` here?

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -64,41 +64,81 @@ jobs:
...
+             echo "${d_tag}registry=${d_registry}" >> $GITHUB_ENV
+             echo "${d_tag}username=${d_username}" >> $GITHUB_ENV
+             echo "${d_tag}password=${d_password}" >> $GITHUB_ENV
 
-    - name: Log in to Docker repository
-      # Only bother if we're going to push the container and we have
-      #  credentials
+    # Only bother logging in to a given site if we're going to push the
+    # container and we have credentials
+    - name: Log in to Docker Hub repository
```
Comment (timj) [inline .github/workflows/build.yaml:109]: It's a shame there is so much code duplication here.

File: Makefile [hunk 4]
```diff
@@ -110,24 +110,37 @@ all: push
 # push assumes that the building user already has docker credentials
 #  to push to whatever the target repository (specified in $(image)) may be.
```
Comment (timj) [inline Makefile:111]: Extra space in start of comment.

File: Makefile [hunk 5]
```diff
@@ -110,24 +110,37 @@ all: push
...
-	$(DOCKER) push $(image):$(version)
-ifneq ($(ltype),)
-	$(DOCKER) tag $(image):$(version) $(image):$(ltype)
-	$(DOCKER) push $(image):$(ltype)
-endif
-ifneq ($(latest),)
-	$(DOCKER) tag $(image):$(version) $(image):$(latest)
-	$(DOCKER) push $(image):$(latest)
-endif
+	img=$$(echo $(image) | cut -d ',' -f 1) && \
```
Comment (timj) [inline Makefile:113]: It would help me if I knew what form `$(image)` had.

**Review by timj — APPROVED** [review]
It seems fine although I don't know much about the content of some of these variables that are being hacked up in the same way in multiple places.

The code duplication is quite distressing but makefiles and YAML don't make it easy.

File: .github/workflows/build.yaml (hunk 3 above)
Comment (athornton) [reply .github/workflows/build.yaml:109]: In theory I could do it with a matrix, but that would be even more confusing.

File: .github/workflows/build.yaml (hunk 2 above)
Comment (athornton) [reply .github/workflows/build.yaml:96]: Shouldn't be a fallthrough.  Strange that it's actually working to push to Google, actually.

File: Makefile (hunk 4 above)
Comment (athornton) [reply Makefile:111]: Consistent with how I've done it in the rest of the file, which is an outdent for the first sentence of each paragraph.

File: Makefile (hunk 5 above)
Comment (athornton) [reply Makefile:115]: Comment added to tell you.

File: .github/workflows/build.yaml (hunk 1 above)
Comment (athornton) [reply .github/workflows/build.yaml]: Done.

**Pull request merged**